pcp/PCPDynamicColumn.c

```diff
@@ -0,0 +1,336 @@
+   PCPDynamicColumnMetric* metric;
+   char* p;
+
+   if ((p = strchr(key, '.')) == NULL)
+      return;
+   *p++ = '\0'; /* end the name, p is now the attribute, e.g. 'label' */
+
+   if (String_eq(p, "metric")) {
+      /* lookup a dynamic metric with this name, else create */
+      metric = PCPDynamicColumn_lookupMetric(columns, column, key);
+
+      /* use derived metrics in dynamic columns for simplicity */
+      char* error;
+      if (pmRegisterDerivedMetric(metric->name, value, &error) < 0) {
+         char* note;
+         xAsprintf(&note,
+                   "%s: failed to parse expression in %s at line %u\n%s\n",
+                   pmGetProgname(), path, line, error);
+         free(error);
+         errno = EINVAL;
+         CRT_fatalError(note);
+         free(note);
+      }
+   } else {
+      /* this is a property of a dynamic metric - the metric expression */
+      /* may not have been observed yet - i.e. we allow for any ordering */
+      metric = PCPDynamicColumn_lookupMetric(columns, column, key);
+      if (String_eq(p, "color")) {
+         if (String_eq(value, "gray"))
+             metric->color = DYNAMIC_GRAY;
+         else if (String_eq(value, "darkgray"))
+             metric->color = DYNAMIC_DARKGRAY;
+         else if (String_eq(value, "red"))
+             metric->color = DYNAMIC_RED;
+         else if (String_eq(value, "green"))
+             metric->color = DYNAMIC_GREEN;
+         else if (String_eq(value, "blue"))
+             metric->color = DYNAMIC_BLUE;
+         else if (String_eq(value, "cyan"))
+             metric->color = DYNAMIC_CYAN;
+         else if (String_eq(value, "magenta"))
+             metric->color = DYNAMIC_MAGENTA;
+         else if (String_eq(value, "yellow"))
+             metric->color = DYNAMIC_YELLOW;
+         else if (String_eq(value, "white"))
+             metric->color = DYNAMIC_WHITE;
+      } else if (String_eq(p, "label")) {
+         char* label = String_cat(value, ": ");
+         free_and_xStrdup(&metric->label, label);
+         free(label);
+      } else if (String_eq(p, "suffix")) {
+         free_and_xStrdup(&metric->suffix, value);
+      }
+   }
+}
+
+// Ensure a valid name for use in a PCP metric name and in htoprc
+static void PCPDynamicColumn_validateColumnName(char* key, const char* path, unsigned int line) {
+   char* p = key;
+   char* end = strrchr(key, ']');
+
+   if (end) {
+      *end = '\0';
+   } else {
+      char* note;
+      xAsprintf(&note,
+                "%s: no closing brace on column....... name at %s line %u\n\"%s\"",
+                pmGetProgname(), path, line, key);
+      errno = EINVAL;
+      CRT_fatalError(note);
+      free(note);
+   }
+
+   while (*p) {
+      if (p == key) {
+         if (!isalpha(*p) && *p != '_')
+            break;
+      } else {
+         if (!isalnum(*p) && *p != '_')
+            break;
```

Comment:
Maybe also allow ``-``?

Reply:
Hyphen (minus) is the syntax used for subtraction in the PCP derived metrics specification, so we must exclude it here when checking name validity.